initial commit of HH\Lib\Str\Utf8 library

ssandler opened the pull request:
## Purpose

Add mb_* string functions to HSL to handle multi-byte strings. We use these quite a lot, especially `mb_strlen` and `mb_strpos`.

## Decisions Made
- Mirror `HH\Lib\Str` function signatures and names
- Be opinionated about encoding, supporting only UTF-8. Do not rely on `mb_internal_encoding`
- Use a sub-namespace of `HH\Lib\Str`, thinking that we'll also add other subnamespaces like `Grapheme` in another PR
- Do not duplicate `Str\` functions into this namespace which do not count bytes or characters, like `contains` and `join` and `ends_with`. Only functions where the distinction between a byte and a multi-byte character matters are implemented differently here. Perhaps even if we eventually move `Str\length` into `Str\Byte\length`, we'll want to keep the functions whose behavior doesn't depend on the interpretation of length/character count in the `Str` namespace?
- Don't implement every mb_* string function, just the ones we find most useful. The most obscure one in this list is probably `convert_kana`, we find it very useful to normalize half-width and full-width Kana for comparing Japanese strings
- `is_valid` replicates a check we occasionally do to test if a string is UTF-8
- Add an enum listing encodings, and support converting from those encodings to UTF-8, or from UTF-8 to those encodings. This is a trimmed down list of everything I got from `mb_list_encodings()`, cross referenced with wikipedia to get a list of commonly used encodings. I would be happy to trim this down more or add more encodings, I don't have strong opinions about this list.


Hoping to get feedback on the overall approach here. Once we agree on the namespaces and general set of functionality, I'll add unit tests to this PR.

kmeht commented on src/str/encoding.php:
```diff
@@ -0,0 +1,60 @@
+<?hh // strict
+/*
+ *  Copyright (c) 2004_present, Facebook, Inc.
```
typo: underscore

ssandler commented:
I guess the Travis build needs to run on nightlies so that it doesn't get a parse error on the new regex syntax? :-D seems unrelated to my PR though.

kmeht commented on src/str/utf8/introspect.php:
```diff
@@ -0,0 +1,112 @@
+ * Returns the length of the given string.
+ * A multi-byte character is counted as 1.
+ *
+ * Previously known in PHP as `mb_strlen`.
```
nit: Consistent between "Previously known in PHP as *" vs. "Previously known as * in PHP". We've gotten this wrong too, lol.

kmeht commented on src/str/utf8/introspect.php:
```diff
@@ -0,0 +1,112 @@
+ * of the string. If the offset is out-of-bounds, a ViolationException will be
+ * thrown.
+ *
+ * - To simply check if the haystack contains the needle, see `Str\contains()`.
```
Reference `contains_ci`

kmeht commented on src/str/utf8/introspect.php:
```diff
@@ -0,0 +1,112 @@
+ * Determine whether a string of unknown encoding is valid UTF-8
+ */
+<<__RxLocal>>
+function is_valid(string $string): bool {
```
`is_utf8`?

ssandler commented:
## Updates

- addressed feedback Kunal posted
- changed namespaces to `HH\Lib\Experimental\Str\*` as Fred recommended in Slack
- added `Grapheme\` functions
- added unit tests

Note that the Grapheme namespace has fewer functions than the Utf8 one. For example there's no `Grapheme\uppercase()`. This is because if that function existed it would be exactly the same as `Utf8\uppercase()`.

If you think it's cleaner I could duplicate the functions into the Utf8 and Grapheme namespaces anyway even if they have no behavior differences. Or perhaps the docs for this can call out the use cases for the different string namespaces.

fredemmott commented:
Sorry for the delay here; the test failure was because we hadn't updated the Trusty nightlies for a month, and I spend a few days (a) fixing that (b) adding more thorough monitoring of our nightlies.

Re-triggered the test run here, and will look back later today or monday.

fredemmott commented on src/str/utf8/transform.php:
```diff
@@ -0,0 +1,102 @@
+ *          Use with "K","H"
+ */
+<<__RxLocal>>
+function convert_kana(string $string, string $options): string {
```
what are the combining rules? Can they be expressed in Hack's type system somehow? (well, without FormatString :p)

Maybe a shape for vec<SomeEnum> makes more sense here ?

ssandler replied:
Updated to use a shape

fredemmott commented on src/str/grapheme/select.php:
```diff
@@ -0,0 +1,62 @@
+ * $start position (in bytes).
+ *
+ * Returns a tuple containing the extracted grapheme(s) and the next byte position to use
+ * for subsequent calls. If the end of the string is reached, null will be returned.
```
`(null, ???)` will be returned

- what will the second part be?
- would `?(string, int)` be better?

fredemmott commented on src/str/grapheme/select.php:
```diff
@@ -0,0 +1,62 @@
+ * Function to extract a sequence of default grapheme clusters from a text buffer,
+ * which must be encoded in UTF-8.
+ *
+ * The $size determines the number of graphemes to extract, starting at the
```
backticks around size, or use `@param`

fredemmott commented on tests/utf8/Utf8IntrospectTest.php:
```diff
@@ -0,0 +1,140 @@
+use function Facebook\FBExpect\expect;
+// @oss-disable: use InvariantViolationException as InvariantException;
+
+final class Utf8IntrospectTest extends PHPUnit_Framework_TestCase {
```
These need to use HackTest now

fredemmott commented on tests/utf8/Utf8IntrospectTest.php:
```diff
@@ -0,0 +1,140 @@
+
+final class Utf8IntrospectTest extends PHPUnit_Framework_TestCase {
+
+  public static function provideLength(): varray<mixed> {
```
`varray<(string, int)>` - though you should be able to use `vec<(string ,int)>` with HackTest

fredemmott commented on tests/utf8/Utf8IntrospectTest.php:
```diff
@@ -0,0 +1,140 @@
+    expect(Utf8\length($string))->toBeSame($expected);
+  }
+
+  public static function provideSearch(): varray<mixed> {
```
similar

fredemmott commented:
Also:
- unsure if we should add a generic wrapper around mb_detect_encoding
- probalby want a Utf8\validate() if not

fredemmott commented:
Oh, you have is_valid but I missed it, sorry :)

fredemmott commented:
> Do not duplicate Str\ functions into this namespace which do not count bytes or characters, like contains and join and ends_with. Only functions where the distinction between a byte and a multi-byte character matters are implemented differently here. Perhaps even if we eventually move Str\length into Str\Byte\length, we'll want to keep the functions whose behavior doesn't depend on the interpretation of length/character count in the Str namespace?

I think this is the best choice for an initial pull request to keep it reviewable - but I pretty strongly feel this is not the right choice long-term, as it depends on every engineer remembering (and thinking about) which to use; in particular, it's common to forget that case-insensitive functions (e.g. `Str\contains_ci`) should not be used. We're generally against duplication, but I think the tradeoff for consistency and easy-to-use-safely wins here. Not sure how @kmeht feels about this though

kmeht commented:
This looks really great! Unless @fredemmott has any other comments, I'm in favor of merging. Random comments:
- HackTestCase was recently renamed to HackTest, we may need to make a new release for that if this is going to work internally.
- It'd be nice to build a native alternative to `convert_kana` that can handle the shape directly, but we'll cross that bridge when we get there.

fredemmott commented:
> HackTestCase was recently renamed to HackTest, we may need to make a new release for that if this is going to work internally.

Yeah, this needs to happen, and I need to update HHAST. Hopefully today.

fredemmott commented:
So, that form isn't going to quite be right once we have a new release up thati ncludes https://github.com/hhvm/hacktest/commit/a3fdc66f0fc002fdff6168766f6c79dc345fabeb

ssandler commented:
Let me know when/if to change the name of the class on this PR (or do we want to do it in a follow up)?

I can also make a follow up PR that copies all the other Str functions into Utf8 and Grapheme as we discussed.

ssandler commented:
Checking in on this, has there been feedback internally on the imported PR?

ssandler commented:
Updated for latest release of HackTest, rebased against origin/master

kmeht commented:
🥇
